src/main/java/com/faforever/api/user/GogService.java

```diff
@@ -0,0 +1,177 @@
+package com.faforever.api.user;
+
+import com.faforever.api.config.FafApiProperties;
+import com.faforever.api.data.domain.User;
+import com.faforever.api.error.ApiException;
+import com.faforever.api.error.ErrorCode;
+import com.google.gson.Gson;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.HttpClientErrorException;
+import org.springframework.web.client.RestTemplate;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+@Service
+@Slf4j
+@RequiredArgsConstructor
+public class GogService {
+  // TODO: what characters to allow?
+  private static final Pattern GOG_USERNAME_PATTERN = Pattern.compile("[a-z\\d-_!?.]+"); // gets lower cased by controller, gog doesn't care about capitalization, this causes db entries to be unique
```

Comment:
What characters should we allow for GOG usernames? I'd prefer not to just take any string.

Reply:
I tried to figure it out with fake registrations but they only tell you that there are invalid special characters, not which ones.

Reply:
I take it bad. The implementation of JDK is horrible here and doesn't work at all in this case.